src/treemodel/selection.js

```diff
@@ -0,0 +1,261 @@
+/**
+ * @license Copyright (c) 2003-2015, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+CKEDITOR.define( [
+	'treemodel/liverange',
+	'treemodel/attributelist',
+	'emittermixin',
+	'utils',
+	'ckeditorerror'
+], ( LiveRange, AttributeList, EmitterMixin, utils, CKEditorError ) => {
+	/**
+	 * Represents a selection that is made on nodes in {@link treeModel.Document}. Selection instance is
+	 * created by {@link treeModel.Document}. In most scenarios you should not need to create an instance of Selection.
+	 *
+	 * @class treeModel.Selection
+	 */
+	class Selection {
+		/**
+		 * Creates an empty selection.
+		 *
+		 * @constructor
+		 */
+		constructor() {
+			/**
+			 * List of attributes set on current selection.
+			 *
+			 * @private
+			 * @property {treeModel.AttributeList} _attrs
+			 */
+			this._attrs = new AttributeList();
+
+			/**
+			 * Stores all ranges that are selected.
+			 *
+			 * @private
+			 * @property {Array.<LiveRange>}
+			 */
+			this._ranges = [];
+
+			/**
+			 * Specifies whether the last added range was added as a backward or forward range.
+			 *
+			 * @private
+			 * @property {Boolean}
+			 */
+			this._lastRangeBackward = false;
```

Comment:
The biggest problem here is that there's no spec. The [current one](http://w3c.github.io/selection-api/) does specify only single range selections. There was also Aryeh Gregor's work but since it was based on how things work in the browsers it either described FF's or other browser's behaviour :D.

Anyway, the point is that in the future the native selection may have more features than the current one and if there's a an API design that could accommodate these features in the future, then it's the way to go. And I think that the API you proposed is good because so far it works automatically (anchor and focus are calculated on the fly). If we'll need to introduce some property or to allow setting some property manually (so it's not calculated based on the last range), then we can do that by extending the API.